x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/lock_manager_client.ts

```diff
@@ -0,0 +1,346 @@
+    }
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const result = await this.esClient.search<LockDocument>({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { gt: 'now' } } }],
+        },
+      },
+    });
+
+    const hits = result.hits.hits;
+    return hits[0]?._source;
+  }
+
+  public async acquireWithRetry({
+    metadata,
+    ttl,
+    ...retryOptions
+  }: AcquireOptions & pRetry.Options = {}): Promise<boolean> {
+    return pRetry(async () => {
+      const acquired = await this.acquire({ metadata, ttl });
+      if (!acquired) {
+        this.logger.debug(`Lock "${this.lockId}" not available yet.`);
+        throw new Error(`Lock "${this.lockId}" not available yet`);
+      }
+      return acquired;
+    }, retryOptions ?? { forever: true, maxTimeout: 10_000 });
+  }
+
+  public async extendTtl(ttl = 300000): Promise<boolean> {
+    try {
+      await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        script: {
+          lang: 'painless',
+          source: `
+          if (ctx._source.token == params.token) {
+            long now = System.currentTimeMillis();
+            ctx._source.expiresAt = Instant.ofEpochMilli(now + params.ttl).toString();
+          } else {
+            ctx.op = 'noop';
+          }`,
+          params: {
+            ttl,
+            token: this.token,
+          },
+        },
+        refresh: true,
+      });
+      this.logger.debug(`Lock "${this.lockId}" extended ttl with ${prettyMilliseconds(ttl)}.`);
+      return true;
+    } catch (error) {
+      if (isVersionConflictException(error)) {
+        this.logger.debug(`Lock "${this.lockId}" was released concurrently. Not extending TTL.`);
+        return false;
+      }
+
+      this.logger.error(`Failed to extend lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+}
+
+export async function withLock<T>(
+  {
+    esClient,
+    logger,
+    lockId,
+    metadata,
+    ttl = duration(5, 'minutes').asMilliseconds(),
+    waitForLock = false,
+    retryOptions,
+  }: {
+    esClient: ElasticsearchClient;
+    logger: Logger;
+    lockId: LockId;
+    waitForLock?: boolean;
+    retryOptions?: pRetry.Options;
+  } & AcquireOptions,
+  callback: () => Promise<T>
+): Promise<T | undefined> {
+  const lockManager = new LockManager(lockId, esClient, logger);
+  const acquired =
+    waitForLock ?? retryOptions
+      ? await lockManager.acquireWithRetry({ metadata, ttl, ...retryOptions })
+      : await lockManager.acquire({ metadata, ttl });
+
+  if (!acquired) {
+    logger.debug(`Lock "${lockId}" not acquired. Exiting.`);
+    throw new Error(`Lock "${lockId}" not acquired`);
```

Comment:
can you throw a specific error here ? like `LockAcquisitionError`? this makes it easier for consumers to figure out what's happening